File: akka-essentials/Chapter02/FirstAkkaApplicationScala/src/main/scala/akka/first/app/mapreduce/actors/MasterActor.scala
package akka.first.app.mapreduce.actors

import akka.actor.Actor
import akka.actor.Props
import akka.first.app.mapreduce._
import akka.routing.RoundRobinPool

class MasterActor extends Actor {

  val mapActor = context.actorOf(
    Props[MapActor]().withRouter(RoundRobinPool(nrOfInstances = 5)),
    name = "map"
  )
  val reduceActor = context.actorOf(
    Props[ReduceActor]().withRouter(RoundRobinPool(nrOfInstances = 5)),
    name = "reduce"
  )
  val aggregateActor = context.actorOf(
    Props[AggregateActor](),
    name = "aggregate"
  )

  def receive: Receive = {
    case line: String =>
      mapActor ! line
    case mapData: MapData =>
      reduceActor ! mapData
    case reduceData: ReduceData =>
      aggregateActor ! reduceData
    case Result =>
      aggregateActor forward Result
  }

}

File: akka-essentials/Chapter02/FirstAkkaApplication2/src/main/java/akka/first/app/mapreduce/actors/MasterActor.java
package akka.first.app.mapreduce.actors;

import akka.actor.AbstractActor;
import akka.actor.ActorRef;
import akka.actor.Props;
import akka.first.app.mapreduce.messages.MapData;
import akka.first.app.mapreduce.messages.ReduceData;
import akka.first.app.mapreduce.messages.Result;
import akka.routing.RoundRobinPool;

/**
 * The Master actor is responsible for the instantiation of the child actors.
 * The Master actor is the gateway for all messages that are passed on to the
 * other actors, namely the Map actor and Aggregate actor.
 */
public class MasterActor extends AbstractActor {

    ActorRef mapActor = getContext().actorOf(
        Props.create(MapActor.class).withRouter(new RoundRobinPool(5)),
        "map"
    );
    ActorRef reduceActor = getContext().actorOf(
        Props.create(ReduceActor.class).withRouter(new RoundRobinPool(5)),
        "reduce"
    );
    ActorRef aggregateActor = getContext().actorOf(
        Props.create(AggregateActor.class),
        "aggregate"
    );

    public Receive createReceive() {
        return receiveBuilder()
            .match(String.class, word -> {
                mapActor.tell(word, getSelf());
            })
            .match(MapData.class, mapData -> {
                reduceActor.tell(mapData, getSelf());
            })
            .match(ReduceData.class, reduceData -> {
                aggregateActor.tell(reduceData, getSelf());
            })
            .match(Result.class, result -> {
                aggregateActor.forward(result, getContext());
            })
            .match(Object.class, message -> {
                unhandled(message);
            })
            .build();
    }

}
